src/backend/distributed/shardsplit/shardsplit_decoder.c

```diff
@@ -62,6 +80,192 @@ _PG_output_plugin_init(OutputPluginCallbacks *cb)
+{
+	char *replicationSlotName = ctx->slot->data.name.data;
+	bool isCDCTranslated = false;
+	Relation targetRelation;
+
+	/* Check if the replication slot is CITUS_CDC_SLOT*/
+	if (replicationSlotName != NULL &&
+		strncmp(replicationSlotName, CITUS_SHARD_PREFIX_SLOT, strlen(
+					CITUS_SHARD_PREFIX_SLOT)) != 0)
+	{
+		/* Skip publishing changes for system relations in pg_catalog*/
+		if (relation->rd_rel->relnamespace == PG_CATALOG_NAMESPACE)
+		{
+			return true;
+		}
+
+		char *shardRelationName = RelationGetRelationName(relation);
+		uint64 shardId = ExtractShardIdFromTableName(shardRelationName, true);
+		if (shardId != INVALID_SHARD_ID && ShardExists(shardId))
+		{
+			if (ReferenceTableShardId(shardId))
+			{
+				/*For reference tables, publish the changes only from the coordinator node. */
+				if (!IsCoordinator())
+				{
+					return true;
+				}
+			}
+			else
+			{
+				/* try to get the distributed relation id for the shard */
+				Oid distributedRelationId = RelationIdForShard(shardId);
+
+				if (OidIsValid(distributedRelationId))
+				{
+					targetRelation = RelationIdGetRelation(distributedRelationId);
+					handleSchemaChangesInRelation(relation, targetRelation, change);
+					isCDCTranslated = true;
+				}
+			}
+		}
+
+		if (isCDCTranslated)
+		{
+			pgoutputChangeCB(ctx, txn, targetRelation, change);
+			RelationClose(targetRelation);
+		}
+		else
+		{
+			pgoutputChangeCB(ctx, txn, relation, change);
+		}
+		return true;
+	}
+	return false;
+}
+
+
+static HeapTuple
+GetTupleForTargetSchemaForCdc(HeapTuple sourceRelationTuple,
+							  TupleDesc sourceRelDesc,
+							  TupleDesc targetRelDesc)
+{
+	/* Deform the tuple */
+	Datum *oldValues = (Datum *) palloc0(sourceRelDesc->natts * sizeof(Datum));
+	bool *oldNulls = (bool *) palloc0(sourceRelDesc->natts * sizeof(bool));
+	heap_deform_tuple(sourceRelationTuple, sourceRelDesc, oldValues,
+					  oldNulls);
+
+
+	/* Create new tuple by skipping dropped columns */
+	int nextAttributeIndex = 0;
+	Datum *newValues = (Datum *) palloc0(targetRelDesc->natts * sizeof(Datum));
+	bool *newNulls = (bool *) palloc0(targetRelDesc->natts * sizeof(bool));
+	for (int i = 0; i < targetRelDesc->natts; i++)
+	{
+		if (TupleDescAttr(targetRelDesc, i)->attisdropped)
+		{
+			Datum nullDatum = (Datum) 0;
+			newValues[i] = nullDatum;
+			newNulls[i] = true;
+		}
+		else
+		{
+			newValues[i] = oldValues[nextAttributeIndex];
+			newNulls[i] = oldNulls[nextAttributeIndex];
+			nextAttributeIndex++;
+		}
+	}
+
+	HeapTuple targetRelationTuple = heap_form_tuple(targetRelDesc, newValues, newNulls);
+	return targetRelationTuple;
+}
+
+
+static void
+handleSchemaChangesInRelation(Relation relation, Relation targetRelation,
+							  ReorderBufferChange *change)
+{
+	TupleDesc sourceRelationDesc = RelationGetDescr(relation);
+	TupleDesc targetRelationDesc = RelationGetDescr(targetRelation);
+	if (targetRelationDesc->natts > sourceRelationDesc->natts)
```

Comment:
This handles the case where the shard is clean and the shell table is dirty.

Technically, we can also have the inverse, by dropping a column (making the shards dirty) and then doing citus_activate_node (making the shell table clean).